platform/barefoot/sonic-platform-modules-ufispace/s9180-32x/sonic_platform/fan.py

```diff
@@ -0,0 +1,177 @@
+#!/usr/bin/env python
+
+
+try:
+    from sonic_platform_pddf_base.pddf_fan import PddfFan
+except ImportError as e:
+    raise ImportError(str(e) + "- required module not found")
+
+
+class Fan(PddfFan):
+    """PDDF Platform-Specific Fan class"""
+
+    def __init__(self, tray_idx, fan_idx=0, pddf_data=None, pddf_plugin_data=None, is_psu_fan=False, psu_index=0):
+        # idx is 0-based
+        PddfFan.__init__(self, tray_idx, fan_idx, pddf_data, pddf_plugin_data, is_psu_fan, psu_index)
+
+    # Provide the functions/variables below for which implementation is to be overwritten
+    # Since psu_fan airflow direction cant be read from sysfs, it is fixed as 'F2B' or 'intake'
+
+    def get_speed(self):
+        """
+        Retrieves the speed of fan as a percentage of full speed
+
+        Returns:
+            An integer, the percentage of full fan speed, in the range 0 (off)
+                 to 100 (full speed)
+        """
+        speed_percentage = 0
+        if self.is_psu_fan:
+            attr = "psu_fan{}_speed_rpm".format(self.fan_index)
+            device = "PSU{}".format(self.fans_psu_index)
+            max_speed = int(self.plugin_data['PSU']['PSU_FAN_MAX_SPEED'])
+        else:
+            if self.fan_index == 1:
+                pos = "f"
+                max_speed = int(self.plugin_data['FAN']['FAN_F_MAX_SPEED'])
+            else:
+                pos = "r"
+                max_speed = int(self.plugin_data['FAN']['FAN_R_MAX_SPEED'])
+            attr = "fan{}_{}_speed_rpm".format(self.fantray_index, pos)
+            device = "FAN-CTRL"
+
+        output = self.pddf_obj.get_attr_name_output(device, attr)
+        if not output:
+            return speed_percentage
+
+        output['status'] = output['status'].rstrip()
+        if output['status'].isalpha():
+            return speed_percentage
+        else:
+            speed = int(float(output['status']))
+
+        speed_percentage = round((speed*100)/max_speed)
+        return speed_percentage
+
+    def get_speed_rpm(self):
+        """
+        Retrieves the speed of fan in RPM
+
+        Returns:
+            An integer, Speed of fan in RPM
+        """
+        rpm_speed = 0
+        if self.is_psu_fan:
+            attr = "psu_fan{}_speed_rpm".format(self.fan_index)
+            device = "PSU{}".format(self.fans_psu_index)
+        else:
+            if self.fan_index == 1:
+                pos = "f"
+            else:
+                pos = "r"
+            attr = "fan{}_{}_speed_rpm".format(self.fantray_index, pos)
+            device = "FAN-CTRL"
+
+        output = self.pddf_obj.get_attr_name_output(device, attr)
+
+        if output is None:
+            return rpm_speed
+
+        output['status'] = output['status'].rstrip()
+        if output['status'].isalpha():
+            return rpm_speed
+        else:
+            rpm_speed = int(float(output['status']))
+
+        return rpm_speed
+
+    def get_direction(self):
```

Comment:
Can you check this again? 
I think same implementation is present in the pddf_fan.py class.

Reply:
This device is old design and we can't get fan direction information from CPU or BMC. I tried to take implementation from pddf_fan.py class, the direction will be "N/a". So I still need to overwrite the get_direction() function and set the direction to FAN_DIRECTION_EXHAUST to fit our device. I will update the code later.